File: 20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        # if s == ""
        if len(s) == 0:
            return True
        d = {
            ')': '(',
            '}' : '{',
            ']' : '['
        }
        temp = []
        for item in s:
            if item not in d:
                temp.append(item)
            else:
                # example: s = ")"
                if temp == []:
                    return False
                if d[item] == temp[-1]:
                    temp.pop()
                else:
                    return False
        if len(temp) > 0:
            return False
        return True 
# @lc code=end

File: C#/20.valid-parentheses.cs
/*
 * @lc app=leetcode id=20 lang=csharp
 *
 * [20] Valid Parentheses
 */

// @lc code=start
using System.Collections.Generic;
public class Solution {
    public bool IsValid(string s) {
        Dictionary<char, char> parenthses = new Dictionary<char, char>();
        parenthses[')'] = '(';
        parenthses[']'] = '[';
        parenthses['}'] = '{';
        Stack<char> st = new Stack<char>();
        var p = s.ToCharArray();
        foreach(var item in p){
            if(st.Count > 0 && parenthses.ContainsKey(item) && st.Peek() == parenthses[item]){
                st.Pop();
            }
            else{
                st.Push(item);
            }
        }
        return st.Count == 0;
    }
}
// @lc code=end

